# 패키지 import

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# 데이터 불러오기

In [2]:
path = './data/'
train = pd.read_csv(path+'train.csv')
test = pd.read_csv(path+'test.csv')

In [3]:
print(train.shape)
print(test.shape)

(891, 12)
(418, 11)


# 데이터 살펴보기

## 컬럼 정보

- Survived: 생존 여부 / 0 = 사망, 1 = 생존
- Pclass: 티켓 등급 / 1 = 1등급, 2 = 2등급, 3 = 3등급
- Name: 이름 / (last name, Title, first name, middle name) 형식으로 구성
- Sex: 성별
- Age: 나이 / 1살 미만의 경우 분수, 추정 나이면 xx.5
- SibSp: 형제자매, 배우자 수
- Parch: 부모, 자녀 수
- Ticket: 티켓 번호
- Fare: 운임
- Cabin: 객실 번호
- Embarked: 탑승 선착장 / C = Cherbourg, Q = Queenstown, S = Southampton

In [4]:
train.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [5]:
train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [6]:
train.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

train의 Age, Cabin, Embarked와 test의 Age, Fare, Cabin에 결측치가 있다.

In [7]:
train.info()
print("-"*40)
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB
----------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Passenger

In [8]:
train.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

# EDA
각 column과 생존률의 관계를 살펴보자.

In [9]:
combine = [train, test]

## Pclass

티켓 등급별 생존률

In [10]:
pd.pivot_table(data=train, values='Survived', index='Pclass', aggfunc=['mean', 'sum', 'count'])

mean      sum    count
        Survived Survived Survived
Pclass                            
1       0.629630      136      216
2       0.472826       87      184
3       0.242363      119      491

티켓 등급과 셩별별 생존률

1등급과 2등급 티켓의 여성이 압도적으로 많이 살아남은 것을 볼 수 있다.

In [11]:
pd.pivot_table(data=train, values='Survived', index=['Pclass', 'Sex'], aggfunc=['mean', 'sum', 'count'])

mean      sum    count
               Survived Survived Survived
Pclass Sex                               
1      female  0.968085       91       94
       male    0.368852       45      122
2      female  0.921053       70       76
       male    0.157407       17      108
3      female  0.500000       72      144
       male    0.135447       47      347

티켓 등급별 평균 연령(결측치 미포함)

In [12]:
# train.groupby('Pclass')[['Age']].mean()
pd.pivot_table(data=train, values='Age', index='Pclass', aggfunc=['mean', 'count'])

mean count
              Age   Age
Pclass                 
1       38.233441   186
2       29.877630   173
3       25.140620   355

티켓 등급별 평균 운임

In [13]:
train.groupby('Pclass')[['Fare']].mean()

Fare
Pclass           
1       84.154687
2       20.662183
3       13.675550

## Name

Name에서 Title(호칭)을 분리해 Title column으로 만들어줬다.

In [14]:
for df in combine:
    df['Title'] = df['Name'].map(lambda x: x.split(', ')[1].split('. ')[0])
train['Title']

0        Mr
1       Mrs
2      Miss
3       Mrs
4        Mr
       ... 
886     Rev
887    Miss
888    Miss
889      Mr
890      Mr
Name: Title, Length: 891, dtype: object

Title(호칭)별 생존률

In [15]:
pd.pivot_table(data=train, values='Survived', index='Title', aggfunc=['mean', 'sum', 'count']).sort_values(('count', 'Survived'), ascending=False)

mean      sum    count
              Survived Survived Survived
Title                                   
Mr            0.156673       81      517
Miss          0.697802      127      182
Mrs           0.792000       99      125
Master        0.575000       23       40
Dr            0.428571        3        7
Rev           0.000000        0        6
Major         0.500000        1        2
Col           0.500000        1        2
Mlle          1.000000        2        2
Sir           1.000000        1        1
Ms            1.000000        1        1
Capt          0.000000        0        1
Mme           1.000000        1        1
Lady          1.000000        1        1
Jonkheer      0.000000        0        1
Don           0.000000        0        1
the Countess  1.000000        1        1

수가 적은 호칭들은 Rare라는 이름으로 합쳐줬다.

In [16]:
under_2_titles = ['Mlle', 'Major', 'Col', 'the Countess', 'Capt', 'Ms', 'Sir', 'Lady', 'Mme', 'Don', 'Jonkheer', 'Col', 'Rev', 'Ms', 'Dr', 'Dona']
under_2_titles

['Mlle',
 'Major',
 'Col',
 'the Countess',
 'Capt',
 'Ms',
 'Sir',
 'Lady',
 'Mme',
 'Don',
 'Jonkheer',
 'Col',
 'Rev',
 'Ms',
 'Dr',
 'Dona']

In [17]:
for df in combine:
    df['Title'] = df['Title'].replace(under_2_titles, 'Rare')
train['Title'].value_counts()

Mr        517
Miss      182
Mrs       125
Master     40
Rare       27
Name: Title, dtype: int64

타이타닉에서의 Master는 어린 남자 아이를 가리키는 호칭으로 사용 되었다고 한다.

평균 나이 4.5세

In [18]:
train[train.Title == 'Master']['Age'].mean()

4.574166666666667

Master 호칭의 나이별 생존률

In [19]:
pd.pivot_table(data=train[train.Title == 'Master'], values='Survived', index=['Sex', 'Age'], aggfunc=['mean', 'sum', 'count'])

mean      sum    count
           Survived Survived Survived
Sex  Age                             
male 0.42      1.00        1        1
     0.67      1.00        1        1
     0.83      1.00        2        2
     0.92      1.00        1        1
     1.00      0.60        3        5
     2.00      0.25        1        4
     3.00      1.00        4        4
     4.00      0.40        2        5
     6.00      1.00        1        1
     7.00      0.00        0        2
     8.00      0.50        1        2
     9.00      0.50        2        4
     10.00     0.00        0        1
     11.00     0.50        1        2
     12.00     1.00        1        1

각 호칭별 평균 나이

In [20]:
pd.pivot_table(data=train, values='Age', index=['Title'], aggfunc=['mean'])

mean
              Age
Title            
Master   4.574167
Miss    21.773973
Mr      32.368090
Mrs     35.898148
Rare    42.384615

## Sex

남녀 생존률 비교

In [21]:
pd.pivot_table(data=train, values='Survived', index='Sex', aggfunc=['mean', 'sum', 'count'])

mean      sum    count
        Survived Survived Survived
Sex                               
female  0.742038      233      314
male    0.188908      109      577

남녀 평균 나이(나이 결측치 미반영)

In [22]:
train.groupby('Sex')[['Age']].mean()

Age
Sex              
female  27.915709
male    30.726645

남녀 운임 비교

여자의 평균 운임이 확연이 더 높은 것을 알 수 있다. 왜그럴까?

In [23]:
train.groupby('Sex')[['Fare']].mean()

Fare
Sex              
female  44.479818
male    25.523893

티켓 등급이 평균적으로 높아서 그런것일까 하고 티켓 등급별로 비교해봤다.

그러나 같은 티켓 안에서도 운임이 차이나는 것을 확인할 수 있다. 특히 1등급 여성의 티켓이 많이 차이난다.

In [24]:
pd.pivot_table(data=train, values='Fare', index=['Pclass', 'Sex'], aggfunc=['mean', 'count'])

mean count
                     Fare  Fare
Pclass Sex                     
1      female  106.125798    94
       male     67.226127   122
2      female   21.970121    76
       male     19.741782   108
3      female   16.118810   144
       male     12.661633   347

여성의 운임이 남성의 운임보다 비싼 이유는 여자가 남자보다 객실을 더 많이 예약한 것이 아닐까? 라는 가설 검증하기 위해 티켓 등급, 성별별 객실 예약 비율을 알아봤다.

In [25]:
for df in combine:
    df['isCabin'] = ~df['Cabin'].isna()
train['isCabin']

0      False
1       True
2      False
3       True
4      False
       ...  
886    False
887     True
888    False
889     True
890    False
Name: isCabin, Length: 891, dtype: bool

객실의 유무를 떠나, 1등급 여성의 운임이 평균적으로 높은 것을 확인할 수 있다.

In [26]:
pd.pivot_table(data=train, values=['Fare'], index=['Pclass', 'Sex', 'isCabin'], aggfunc=['mean', 'count'])

mean count
                             Fare  Fare
Pclass Sex    isCabin                  
1      female False    137.680769    13
              True     101.061420    81
       male   False     47.559100    27
              True      72.815703    95
2      female False     23.084659    66
              True      14.614170    10
       male   False     19.502328   102
              True      23.812500     6
3      female False     16.173553   138
              True      14.859717     6
       male   False     12.734198   341
              True       8.537500     6

## Age
결측치 177개 존재

In [27]:
train.Age.isna().sum()

177

나이대별 생존률 확인하기

나이대를 나타내는 "AgeRange" column 생성

In [28]:
AgeRange = train.Age//10*10 # 십의자리 내림 꼼수
AgeRange.value_counts().sort_index()

0.0      62
10.0    102
20.0    220
30.0    167
40.0     89
50.0     48
60.0     19
70.0      6
80.0      1
Name: Age, dtype: int64

60대 이상은 표본이 너무 작기 때문에 60으로 합침

In [29]:
AgeRange[AgeRange >= 60.0] = 60.0
AgeRange.value_counts().sort_index()

0.0      62
10.0    102
20.0    220
30.0    167
40.0     89
50.0     48
60.0     26
Name: Age, dtype: int64

df에 AgeRange column 추가

In [30]:
for df in combine:
    AgeRange = df.Age//10*10
    AgeRange[AgeRange >= 60.0] = 60.0
    df['AgeRange'] = AgeRange
train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked Title  isCabin  AgeRange  
0        0         A/5 21171   7.2500   NaN        S    Mr    False      20.0  
1        0          PC 17599  71.2833   C85        C   Mrs     True      30.0  
2        0  STON/O2. 3101282   7.9250   NaN        S  Miss    False      20.0  
3        0            113803  53.1000  C123        S   Mrs     True      30.0  
4        0            373450   8.0500   NaN        S    Mr    False      30.0  
..     ...               ...      ...   ...      ...   ...      ...       ...  
886      0            211536  13.0000   NaN        S  Rare    False      20.0  
887      0            112053  30.0000   B42        S  Miss     True      10.0  
888      2        W./C. 6607  23.4500   NaN        S  Miss    False       NaN  
889      0            111369  30.0000  C148        C    Mr     True      20.0  
890      0            370376   7.7500   NaN        Q    Mr    False      30.0  

[891 rows x 15 columns]

나이대별 생존률

In [31]:
train.groupby('AgeRange')[['Survived']].mean()

Survived
AgeRange          
0.0       0.612903
10.0      0.401961
20.0      0.350000
30.0      0.437126
40.0      0.382022
50.0      0.416667
60.0      0.269231

나이 정보가 있는 사람들의 평균 생존률은 약 40%이다.

In [32]:
train.groupby('AgeRange')['Survived'].mean().mean()

0.4099870952043884

아래는 전체 생존률과 연령대별 생존률의 차이 수치이다.

10대 미만의 생존률이 평균보다 약 20% 높고, 60대 이상의 생존률이 평균보다 약 14% 떨어지는 것을 확인할 수 있다.

In [33]:
(train.groupby('AgeRange')[['Survived']].mean())-(train.groupby('AgeRange')['Survived'].mean().mean())

Survived
AgeRange          
0.0       0.202916
10.0     -0.008026
20.0     -0.059987
30.0      0.027139
40.0     -0.027965
50.0      0.006680
60.0     -0.140756

## SibSp

형제자매와 배우자 수에 따른 생존률

In [34]:
pd.pivot_table(data=train, values='Survived', index='SibSp', aggfunc=['mean', 'sum', 'count'])

mean      sum    count
       Survived Survived Survived
SibSp                            
0      0.345395      210      608
1      0.535885      112      209
2      0.464286       13       28
3      0.250000        4       16
4      0.166667        3       18
5      0.000000        0        5
8      0.000000        0        7

## Parch

부모와 자식의 수에 따른 생존률

In [35]:
pd.pivot_table(data=train, values='Survived', index='Parch', aggfunc=['mean', 'sum', 'count'])

mean      sum    count
       Survived Survived Survived
Parch                            
0      0.343658      233      678
1      0.550847       65      118
2      0.500000       40       80
3      0.600000        3        5
4      0.000000        0        4
5      0.200000        1        5
6      0.000000        0        1

## SibSp + Parch = Family!
SibSp(형제자매, 배우자)와 Parch(부모, 자식)을 합쳐 Family column을 추가했다.

In [36]:
for df in combine:
    df['Family'] = df['SibSp'] + df['Parch']
train['Family'].value_counts()

0     537
1     161
2     102
3      29
5      22
4      15
6      12
10      7
7       6
Name: Family, dtype: int64

가족 수에 따른 생존률

In [37]:
pd.pivot_table(data=train, values='Survived', index='Family', aggfunc=['mean', 'sum', 'count'])

mean      sum    count
        Survived Survived Survived
Family                            
0       0.303538      163      537
1       0.552795       89      161
2       0.578431       59      102
3       0.724138       21       29
4       0.200000        3       15
5       0.136364        3       22
6       0.333333        4       12
7       0.000000        0        6
10      0.000000        0        7

## Ticket

중복되는 티켓들이 몇 가지 있다.

In [38]:
train['Ticket'].value_counts()

347082      7
CA. 2343    7
1601        7
3101295     6
CA 2144     6
           ..
9234        1
19988       1
2693        1
PC 17612    1
370376      1
Name: Ticket, Length: 681, dtype: int64

중복되는 티켓의 개수

In [39]:
(train['Ticket'].value_counts()>1).sum()

134

## Fare

평균 운임은 32.2, 최소 운임은 0, 최대 운임은 512였다.

In [40]:
train.Fare.describe()

count    891.000000
mean      32.204208
std       49.693429
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      512.329200
Name: Fare, dtype: float64

운임이 0인 데이터를 확인해보자.

티켓 등급은 각각 다르고, 전원 남자에 같이 탑승한 가족들은 없었고 전부 S 승선항에서 탑승했다.

생존률은 0.06%로 아주 처참하였다.

In [41]:
train[(train.Fare == 0)]

PassengerId  Survived  Pclass                              Name   Sex  \
179          180         0       3               Leonard, Mr. Lionel  male   
263          264         0       1             Harrison, Mr. William  male   
271          272         1       3      Tornquist, Mr. William Henry  male   
277          278         0       2       Parkes, Mr. Francis "Frank"  male   
302          303         0       3   Johnson, Mr. William Cahoone Jr  male   
413          414         0       2    Cunningham, Mr. Alfred Fleming  male   
466          467         0       2             Campbell, Mr. William  male   
481          482         0       2  Frost, Mr. Anthony Wood "Archie"  male   
597          598         0       3               Johnson, Mr. Alfred  male   
633          634         0       1     Parr, Mr. William Henry Marsh  male   
674          675         0       2        Watson, Mr. Ennis Hastings  male   
732          733         0       2              Knight, Mr. Robert J  male   
806          807         0       1            Andrews, Mr. Thomas Jr  male   
815          816         0       1                  Fry, Mr. Richard  male   
822          823         0       1   Reuchlin, Jonkheer. John George  male   

      Age  SibSp  Parch  Ticket  Fare Cabin Embarked Title  isCabin  AgeRange  \
179  36.0      0      0    LINE   0.0   NaN        S    Mr    False      30.0   
263  40.0      0      0  112059   0.0   B94        S    Mr     True      40.0   
271  25.0      0      0    LINE   0.0   NaN        S    Mr    False      20.0   
277   NaN      0      0  239853   0.0   NaN        S    Mr    False       NaN   
302  19.0      0      0    LINE   0.0   NaN        S    Mr    False      10.0   
413   NaN      0      0  239853   0.0   NaN        S    Mr    False       NaN   
466   NaN      0      0  239853   0.0   NaN        S    Mr    False       NaN   
481   NaN      0      0  239854   0.0   NaN        S    Mr    False       NaN   
597  49.0      0      0    LINE   0.0   NaN        S    Mr    False      40.0   
633   NaN      0      0  112052   0.0   NaN        S    Mr    False       NaN   
674   NaN      0      0  239856   0.0   NaN        S    Mr    False       NaN   
732   NaN      0      0  239855   0.0   NaN        S    Mr    False       NaN   
806  39.0      0      0  112050   0.0   A36        S    Mr     True      30.0   
815   NaN      0      0  112058   0.0  B102        S    Mr     True       NaN   
822  38.0      0      0   19972   0.0   NaN        S  Rare    False      30.0   

     Family  
179       0  
263       0  
271       0  
277       0  
302       0  
413       0  
466       0  
481       0  
597       0  
633       0  
674       0  
732       0  
806       0  
815       0  
822       0

객실 데이터의 유무에 따른 요금을 비교해보았다.

In [42]:
print(f"객실이 NaN인 사람의 요금: {train.Fare[train.Cabin.isna()].mean()}")
print(f"객실이 있는 사람의 요금: {train.Fare[~train.Cabin.isna()].mean()}")

객실이 NaN인 사람의 요금: 19.157325327510915
객실이 있는 사람의 요금: 76.14150392156863


요금의 차이가 확연히 나는 것을 확인할 수 있었다.

그래서 Cabin이 NaN값인 사람들이 데이터가 기록되지 않은 것이 아닌 앉아서 가는 좌석만을 예매한 것이 아닐까? 라는 가설을 세워봤다.

그러나 자료를 조사해본 결과 타이타닉 호는 영국-미국의 장기 항해 배였고 빙산에 충돌하기 전까지 약 4일 이상을 항해했던 것으로 보아 승객의 객실 배정은 필수 요소였던 것으로 보인다.

결과적으로 객실이 NaN인 사람들도 모두 객실을 배정 받았을 것이다.

## Cabin

결측치 687개 존재

In [43]:
train.Cabin.isna().sum()

687

객실번호 맨 앞글자 알파벳은 층을 의미한다고 한다.

그래서 층별 승객 생존률을 알아보기 위해 Cabin 데이터에서 층 정보를 분리해 Floor column을 만들어줬다.

In [44]:
droped_Cabin = train.Cabin.dropna()
droped_Cabin = droped_Cabin.map(lambda x: x[0])
droped_Cabin

1      C
3      C
6      E
10     G
11     C
      ..
871    D
872    B
879    C
887    B
889    C
Name: Cabin, Length: 204, dtype: object

In [45]:
for df in combine:
    droped_Cabin = df.Cabin.dropna()
    droped_Cabin = droped_Cabin.map(lambda x: x[0])
    df['Floor'] = droped_Cabin
train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked Title  isCabin  AgeRange  \
0        0         A/5 21171   7.2500   NaN        S    Mr    False      20.0   
1        0          PC 17599  71.2833   C85        C   Mrs     True      30.0   
2        0  STON/O2. 3101282   7.9250   NaN        S  Miss    False      20.0   
3        0            113803  53.1000  C123        S   Mrs     True      30.0   
4        0            373450   8.0500   NaN        S    Mr    False      30.0   
..     ...               ...      ...   ...      ...   ...      ...       ...   
886      0            211536  13.0000   NaN        S  Rare    False      20.0   
887      0            112053  30.0000   B42        S  Miss     True      10.0   
888      2        W./C. 6607  23.4500   NaN        S  Miss    False       NaN   
889      0            111369  30.0000  C148        C    Mr     True      20.0   
890      0            370376   7.7500   NaN        Q    Mr    False      30.0   

     Family Floor  
0         1   NaN  
1         1     C  
2         0   NaN  
3         1     C  
4         0   NaN  
..      ...   ...  
886       0   NaN  
887       0     B  
888       3   NaN  
889       0     C  
890       0   NaN  

[891 rows x 17 columns]

층별 승객 생존률

In [46]:
pd.pivot_table(data=train, values='Survived', index='Floor', aggfunc=['mean', 'sum', 'count'])

mean      sum    count
       Survived Survived Survived
Floor                            
A      0.466667        7       15
B      0.744681       35       47
C      0.593220       35       59
D      0.757576       25       33
E      0.750000       24       32
F      0.615385        8       13
G      0.500000        2        4
T      0.000000        0        1

티켓 등급별 객실 위치를 파악해보자

In [47]:
train.groupby('Pclass')['Floor'].value_counts()

Pclass  Floor
1       C        59
        B        47
        D        29
        E        25
        A        15
        T         1
2       F         8
        D         4
        E         4
3       F         5
        G         4
        E         3
Name: Floor, dtype: int64

## Embarked

결측치 2개 존재

In [48]:
train.Embarked.isna().sum()

2

승선항별 생존률 분석

In [49]:
pd.pivot_table(data=train, values='Survived', index='Embarked', aggfunc=['mean', 'sum', 'count'], sort=False)

mean      sum    count
          Survived Survived Survived
Embarked                            
S         0.336957      217      644
C         0.553571       93      168
Q         0.389610       30       77

승선항별 평균 티켓 등급과 요금

C는 평균 티켓 등굽과 운임이 높은 것을 보아 부자동네 근처일 것이라고 생각할 수 있다.

Q는 평균 티켓 등급이 2.9로 일반 서민 동네 근처라고 추론해볼 수 있다.

In [50]:
train.groupby('Embarked')[['Pclass', 'Fare']].mean()

Pclass       Fare
Embarked                     
C         1.886905  59.954144
Q         2.909091  13.276030
S         2.350932  27.079812

# 전처리

## 데이터 split

train과 test는 같은 분포에서 추출되기 때문에 train에서 적용한 값을 적용하자.

In [51]:
# from sklearn.model_selection import train_test_split

# train_test_split(train)

In [52]:
print(train.shape, test.shape)

(891, 17) (418, 16)


## 사용하지 않는 열 삭제

PassengerId, Name, Ticket, Cabin(test의 PassengerId는 제출시 사용되므로 삭제하지 않기)

In [53]:
train = train.drop(['PassengerId', 'Name', 'Ticket', 'Cabin'], axis=1)
test = test.drop(['Name', 'Ticket', 'Cabin'], axis=1)

In [54]:
train

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked Title  \
0           0       3    male  22.0      1      0   7.2500        S    Mr   
1           1       1  female  38.0      1      0  71.2833        C   Mrs   
2           1       3  female  26.0      0      0   7.9250        S  Miss   
3           1       1  female  35.0      1      0  53.1000        S   Mrs   
4           0       3    male  35.0      0      0   8.0500        S    Mr   
..        ...     ...     ...   ...    ...    ...      ...      ...   ...   
886         0       2    male  27.0      0      0  13.0000        S  Rare   
887         1       1  female  19.0      0      0  30.0000        S  Miss   
888         0       3  female   NaN      1      2  23.4500        S  Miss   
889         1       1    male  26.0      0      0  30.0000        C    Mr   
890         0       3    male  32.0      0      0   7.7500        Q    Mr   

     isCabin  AgeRange  Family Floor  
0      False      20.0       1   NaN  
1       True      30.0       1     C  
2      False      20.0       0   NaN  
3       True      30.0       1     C  
4      False      30.0       0   NaN  
..       ...       ...     ...   ...  
886    False      20.0       0   NaN  
887     True      10.0       0     B  
888    False       NaN       3   NaN  
889     True      20.0       0     C  
890    False      30.0       0   NaN  

[891 rows x 13 columns]

## 결측치 채우기

In [55]:
train.isna().sum()

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Fare          0
Embarked      2
Title         0
isCabin       0
AgeRange    177
Family        0
Floor       687
dtype: int64

In [56]:
test.isna().sum()

PassengerId      0
Pclass           0
Sex              0
Age             86
SibSp            0
Parch            0
Fare             1
Embarked         0
Title            0
isCabin          0
AgeRange        86
Family           0
Floor          327
dtype: int64

### Age, AgeRange

Age의 결측치를 채울 때 Pclass와 Title을 고려하기로 했다. Title이 Master과 Rare인 사람은 표본 수가 너무 적어 Pclass별로 나누기 전의 평균으로 채워주고, 나머지는 Pclass를 고려해 평균으로 채워주겠다.

AgeRange도 채운 값에따라 갱신해주자.

In [57]:
pd.pivot_table(data=train, values='Age', index=['Pclass', 'Title'], aggfunc=['mean', 'count'])

mean count
                     Age   Age
Pclass Title                  
1      Master   5.306667     3
       Miss    30.000000    45
       Mr      41.580460    87
       Mrs     40.882353    34
       Rare    43.411765    17
2      Master   2.258889     9
       Miss    22.390625    32
       Mr      32.768293    82
       Mrs     33.682927    41
       Rare    40.444444     9
3      Master   5.350833    24
       Miss    16.123188    69
       Mr      28.724891   229
       Mrs     33.515152    33

master 채우기

In [58]:
master_age = train[train['Title'] == 'Master']['Age'].mean()
master_age

4.574166666666667

In [59]:
train.loc[train['Title'] == 'Master', 'Age'] = train.loc[train['Title'] == 'Master', 'Age'].fillna(master_age)
test.loc[test['Title'] == 'Master', 'Age'] = test.loc[test['Title'] == 'Master', 'Age'].fillna(master_age)

Rare 채우기

In [60]:
rare_age = train[train['Title'] == 'Rare']['Age'].mean()
rare_age

42.38461538461539

In [61]:
train.loc[train['Title'] == 'Rare', 'Age'] = train.loc[train['Title'] == 'Rare', 'Age'].fillna(rare_age)
test.loc[test['Title'] == 'Rare', 'Age'] = test.loc[test['Title'] == 'Rare', 'Age'].fillna(rare_age)

나머지 채우기

In [62]:
arr = [[1,2,3], ['Mr','Miss','Mrs']]

In [63]:
for i in arr[0]:
    for j in arr[1]:
        age_mean = train.loc[(train.Pclass == i) & (train.Title == j), 'Age'].mean()
        train.loc[(train.Pclass == i) & (train.Title == j), 'Age'] = train.loc[(train.Pclass == i) & (train.Title == j), 'Age'].fillna(age_mean)


In [64]:
for i in arr[0]:
    for j in arr[1]:
        age_mean = test.loc[(test.Pclass == i) & (test.Title == j), 'Age'].mean()
        test.loc[(test.Pclass == i) & (test.Title == j), 'Age'] = test.loc[(test.Pclass == i) & (test.Title == j), 'Age'].fillna(age_mean)


AgeRange 채우기

In [65]:
AgeRange = train.Age//10*10
AgeRange[AgeRange >= 60.0] = 60.0
train['AgeRange'] = AgeRange

AgeRange = test.Age//10*10
AgeRange[AgeRange >= 60.0] = 60.0
test['AgeRange'] = AgeRange

## Embarked

티켓 번호가 같은 것을 보아 같은 승선항에서 탑승한 것으로 보인다.

In [66]:
train[train.Embarked.isna()]

Survived  Pclass     Sex   Age  SibSp  Parch  Fare Embarked Title  \
61          1       1  female  38.0      0      0  80.0      NaN  Miss   
829         1       1  female  62.0      0      0  80.0      NaN   Mrs   

     isCabin  AgeRange  Family Floor  
61      True      30.0       0     B  
829     True      60.0       0     B

다른 1등급 여성의 Embarked를 확인한 결과 S와 C의 수치가 비슷하기 때문에 최빈값인 S로 설정해줬다.

In [67]:
train[(train.Pclass == 1)&(train.Sex == 'female')]['Embarked'].value_counts()

S    48
C    43
Q     1
Name: Embarked, dtype: int64

In [68]:
train.Embarked = train.Embarked.fillna('S')
train.Embarked.isna().sum()

0

### Floor

Floor는 결측치의 비율이 절반 이상이기 때문에 삭제하는게 낫다고 판단했다.

In [69]:
train = train.drop(['Floor'], axis=1)
test = test.drop(['Floor'], axis=1)

### test의 Fare

결측치가 한 개 이기 때문에 운임에 영향을 주는 요인인 Pclass, Sex, Embarked, Family를 고려한 평균값으로 채워줬다.

In [70]:
test[test.Fare.isna()]

PassengerId  Pclass   Sex   Age  SibSp  Parch  Fare Embarked Title  \
152         1044       3  male  60.5      0      0   NaN        S    Mr   

     isCabin  AgeRange  Family  
152    False      60.0       0

In [71]:
test.Fare = test.Fare.fillna(train[(train.Pclass == 3)&(train.Sex == 'male')&(train.Embarked == 'S')&(train.Family == 0)]['Fare'].mean())
test.Fare.isna().sum()

0

In [72]:
train.isna().sum()

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    0
Title       0
isCabin     0
AgeRange    0
Family      0
dtype: int64

## 범주형 변수 수치형으로 변경

Sex, Embarked, Title, isCabin

In [73]:
train

Survived  Pclass     Sex        Age  SibSp  Parch     Fare Embarked  \
0           0       3    male  22.000000      1      0   7.2500        S   
1           1       1  female  38.000000      1      0  71.2833        C   
2           1       3  female  26.000000      0      0   7.9250        S   
3           1       1  female  35.000000      1      0  53.1000        S   
4           0       3    male  35.000000      0      0   8.0500        S   
..        ...     ...     ...        ...    ...    ...      ...      ...   
886         0       2    male  27.000000      0      0  13.0000        S   
887         1       1  female  19.000000      0      0  30.0000        S   
888         0       3  female  16.123188      1      2  23.4500        S   
889         1       1    male  26.000000      0      0  30.0000        C   
890         0       3    male  32.000000      0      0   7.7500        Q   

    Title  isCabin  AgeRange  Family  
0      Mr    False      20.0       1  
1     Mrs     True      30.0       1  
2    Miss    False      20.0       0  
3     Mrs     True      30.0       1  
4      Mr    False      30.0       0  
..    ...      ...       ...     ...  
886  Rare    False      20.0       0  
887  Miss     True      10.0       0  
888  Miss    False      10.0       3  
889    Mr     True      20.0       0  
890    Mr    False      30.0       0  

[891 rows x 12 columns]

In [74]:
train.Sex = train.Sex.map({'male':0, 'female':1})
test.Sex = test.Sex.map({'male':0, 'female':1})

In [75]:
train.Embarked = train.Embarked.map({'S':0, 'C':1, 'Q':2})
test.Embarked = test.Embarked.map({'S':0, 'C':1, 'Q':2})

In [76]:
train.Title = train.Title.map({'Mr':0, 'Miss':1, 'Mrs':2, 'Master':3, 'Rare':4})
test.Title = test.Title.map({'Mr':0, 'Miss':1, 'Mrs':2, 'Master':3, 'Rare':4})

In [77]:
train.isCabin = train.isCabin.map({False:0, True:1})
test.isCabin = test.isCabin.map({False:0, True:1})

In [78]:
train

Survived  Pclass  Sex        Age  SibSp  Parch     Fare  Embarked  Title  \
0           0       3    0  22.000000      1      0   7.2500         0      0   
1           1       1    1  38.000000      1      0  71.2833         1      2   
2           1       3    1  26.000000      0      0   7.9250         0      1   
3           1       1    1  35.000000      1      0  53.1000         0      2   
4           0       3    0  35.000000      0      0   8.0500         0      0   
..        ...     ...  ...        ...    ...    ...      ...       ...    ...   
886         0       2    0  27.000000      0      0  13.0000         0      4   
887         1       1    1  19.000000      0      0  30.0000         0      1   
888         0       3    1  16.123188      1      2  23.4500         0      1   
889         1       1    0  26.000000      0      0  30.0000         1      0   
890         0       3    0  32.000000      0      0   7.7500         2      0   

     isCabin  AgeRange  Family  
0          0      20.0       1  
1          1      30.0       1  
2          0      20.0       0  
3          1      30.0       1  
4          0      30.0       0  
..       ...       ...     ...  
886        0      20.0       0  
887        1      10.0       0  
888        0      10.0       3  
889        1      20.0       0  
890        0      30.0       0  

[891 rows x 12 columns]

# 모델링

In [79]:
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from catboost import CatBoostClassifier

from sklearn.metrics import accuracy_score

In [80]:
train_X, train_y = train.drop('Survived', axis=1), train.Survived
test_X = test.drop('PassengerId', axis=1)

In [81]:
model_lr = LogisticRegression()
model_lr.fit(train_X, train_y)
train_pred_lr = model_lr.predict(train_X)
test_pred_lr = model_lr.predict(test_X)
accuracy_score(train_y, train_pred_lr)

c:\Users\sunse\miniconda3\envs\coding\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.8159371492704826

In [82]:
model_knn = KNeighborsClassifier()
model_knn.fit(train_X, train_y)
train_pred_knn = model_knn.predict(train_X)
test_pred_knn = model_knn.predict(test_X)
accuracy_score(train_y, train_pred_knn)

0.8159371492704826

In [83]:
model_svc = SVC()
model_svc.fit(train_X, train_y)
train_pred_svc = model_svc.predict(train_X)
test_pred_svc = model_svc.predict(test_X)
accuracy_score(train_y, train_pred_svc)

0.7081930415263749

In [84]:
model_nb = GaussianNB()
model_nb.fit(train_X, train_y)
train_pred_nb = model_nb.predict(train_X)
test_pred_nb = model_nb.predict(test_X)
accuracy_score(train_y, train_pred_nb)

0.7856341189674523

In [85]:
model_tree = DecisionTreeClassifier()
model_tree.fit(train_X, train_y)
train_pred_tree = model_tree.predict(train_X)
test_pred_tree = model_tree.predict(test_X)
accuracy_score(train_y, train_pred_tree)

0.9876543209876543

In [86]:
model_xgb = XGBClassifier()
model_xgb.fit(train_X, train_y)
train_pred_xgb = model_xgb.predict(train_X)
test_pred_xgb = model_xgb.predict(test_X)
accuracy_score(train_y, train_pred_xgb)

0.9764309764309764

In [87]:
model_lgbm = LGBMClassifier()
model_lgbm.fit(train_X, train_y)
train_pred_lgbm = model_lgbm.predict(train_X)
test_pred_lgbm = model_lgbm.predict(test_X)
accuracy_score(train_y, train_pred_lgbm)

0.9629629629629629

In [88]:
model_cat = CatBoostClassifier()
model_cat.fit(train_X, train_y, verbose=False)
train_pred_cat = model_cat.predict(train_X)
test_pred_cat = model_cat.predict(test_X)
accuracy_score(train_y, train_pred_cat)

c:\Users\sunse\miniconda3\envs\coding\lib\site-packages\catboost\core.py:1411: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  self._init_pool(data, label, cat_features, text_features, embedding_features, embedding_features_data, pairs, weight,


0.9090909090909091

In [89]:
test_pred_cat

array([0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1,
       1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1,
       1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0,
       0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0,
       0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0,
       1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1,
       0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0,

# 제출

In [96]:
preds = {'tree':test_pred_tree, 'xgb':test_pred_xgb, 'lgbm':test_pred_lgbm, 'cat':test_pred_cat}

for name, pred in preds.items():
    submission = pd.DataFrame({
            "PassengerId": test["PassengerId"],
            "Survived": pred
        })
    submission.to_csv(f'{name}.csv', index=False)

# Reference
- https://ko.wikipedia.org/wiki/RMS_%ED%83%80%EC%9D%B4%ED%83%80%EB%8B%89
- https://www.kaggle.com/code/startupsci/titanic-data-science-solutions